Curve Manager Round 2

Myoldmopar opened the pull request:
Pull request overview
---------------------
Per the discussion on the call, this starts formalizing the API (number of indep vars).
  - Added individual overloads to `CurveValue` for each number of args.  Each overload only implements the curve types for that number of args, and if it falls through, it will call a fallBack function
  - In client code which is known to have variable number of args, it adds a switch block to catch each case and call the right overload.
    - A default: block will allow user files to continue, but will issue an info message in the err file asking the user to report the corner case
  - Tagged the calls to CheckCurveDims that have multiple possible args with `// MULTIPLECURVEDIMS` so they could be easier to find until later
  - Fleshed out the full possible set of calls to CurveValue in the EMS `@CurveValue` function.  It has to be able to catch each possible case.
  - Fixed some unit tests where curves were being constructed manually, but the `numDims` was not being set.  This was allowing some curve evaluation to fall through silently.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

Myoldmopar committed:
Broke curve funcs by num indep vars

Myoldmopar committed:
Eliminating intermediate functions, trying to expunge optionals from most of curve manager

Myoldmopar committed:
Tag all CheckCurveDim calls for multiple dimensions with MULTIPLECURVEDIMS

Myoldmopar committed:
Add fallback CurveValue function to allow simulations to pass

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into CurveManagerCleanDec2022

Myoldmopar committed:
Add fallback message to catch corner cases

Myoldmopar committed:
Add switch blocks in known client code with variable num args

Myoldmopar commented:
Some questions and future thinking:
 - What value do curve input/output variables hold?  Curves can be reused by any number of components, and when the output processor grabs the value, it's just grabbing the last evaluated call.  I understand EMS has some value.  You can create a unique curve for a chiller, then use EMS to override the curve output and essentially create your own custom chiller logic in a tiny little EMS function.  But output variables...I don't see it.
 - Next step is to break the curves into individual classes for each formulation.  However, as it is right now, we may have to keep a bit of extra bulk in there for handing the 'fallback' curve cases and such.  It may be better to pause with the curve manager after this, let a release identify the corner cases, rip off those band-aids, and push the refactor again after that.  I'm fine either way, just thinking out loud.

amirroth commented:
This is proceeding quite nicely. I have several thoughts here:
- Even after you "object orient" the Curve objects themselves, you will probably want to retain the static function Curve API for EMS and other API-ish purposes. Yes?
- You can push the refactor along other "dimensions" (pun not intended) while keeping the CurveValue cascade logic as is. You are right that it probably makes sense to see how it behaves "in the field" before you attempt to simplify it, decentralize it, or do something else to it. But that is independent of object orientation or other cleanups. That logic "will travel".
- I am not sure what your first question is about, but it sounds like a static/dynamic instance issue. Meaning, you have one static instance of a curve, but multiple logical dynamic users of it and because there is only one copy of the dynamic state, it is the copy of the most recent user. This is a design flaw. You are seeing some of the manifestations of it in reporting and output variables, but the real manifestation will come if EnergyPlus is every internally multi-threaded because this situation is not thread safe. This is not something we have to fix immediately, because EnergyPlus is not internally multi-threaded right now. However, if I understand the description of the problem correctly, then that is what's going on here and the "right" solution is to split the static and dynamic instance state and give each user of the curve its own dynamic instance state, perhaps by turning it into a parameter, or to get rid of dynamic instance state entirely.

amirroth commented:
When I said "refactoring along other dimensions", here is what I mean specifically:
- CurveBase and CurveChildren objects, you already know about this.
- epJSON input API rather than getObjectItem.
- Delete Get/Set APIs. Just access the object directly. You don't need an API to get the name of object.
- Refactor "check" functions into separate check functions (or not) and error functions. Take  `CheckCurveDims` as an example.

```cpp
Curve::CheckCurveDims(
   state,                                                                                        
   state.dataHVACVarRefFlow->VRF(VRFNum).CoolCapFT, 
   {2},                          
   RoutineName,                                                                   
   cCurrentModuleObject,                                                           
   state.dataHVACVarRefFlow->VRF(VRFNum).Name,                                 
   cAlphaFieldNames(3));                          
```

- You are calling a function with seven parameters, one of which is a vector which you are constructing for the sole purpose of calling this function, and four are only used in the rare case of an error. 99.9% of the time, all of this rigamarole exists so that you can confirm `thisCurve.numDims == 2`. There is a better way.

```cpp
if (thisCurve.numDims != 2) {
   ErrorsFound = true;
   Curve::ShowNumDimsError(state, RoutineName, CurrentModuleObject, thisVRF.name, cAlphaFieldNames(3), thisCurveName, "2");
}
```

- Furthermore, given that `RoutineName`, `CurrentModuleObject`, and `x.Name` are going to be parameters to essentially every error function, you may as well combine them into a `struct`.

```cpp
struct ShowErrorParams { std::string_view routineName, objectTypeName, objectName; }
```

- That's all for now.

Myoldmopar committed:
Fix uninitialized variable usage in WAHPSimple

Myoldmopar committed:
Clang format

Myoldmopar commented:
@amirroth, thanks for the comments:

> Even after you "object orient" the Curve objects themselves, you will probably want to retain the static function Curve API for EMS and other API-ish purposes. Yes?

I think for the time being, yeah, we'll probably keep it around.  I'd love to get to a point where don't need it, but it's probably a few steps too far in the future.

> You can push the refactor along other "dimensions" (pun not intended) while keeping the CurveValue cascade logic as is.

Got it, and yeah I'll try to make those changes you suggested in the subsequent comment.

> I am not sure what your first question is about, but it sounds like a static/dynamic instance issue.

Well, the problem is that 10 different heat pumps can share the same curve object by index.  Curves have output variables for the input and output values of the curves.  These were added to help debug problems, but I think they have the potential for more confusion.  You have no idea which of the heat pumps was called last with its own input values and curve output.  I would love to just see them go away...we should have component models register their own meaningful values as output variables, not the generic curve manager.

And on EMS, I do get it's a quick way to allow for an override, but really, we should just make sure that individual component models have the actuators they need to run so that the curve doesn't need an EMS override.  

If we eliminated both of those aspects from curve manager, it would essentially become a flat set of equations.  I'm not proposing we take that step now (or ever), just explaining what's going on and why.

Myoldmopar commented:
One quick question about the child classes.  Right now I've broken up the CurveValue() function into 6 overloads, each with a specific number of independent variable arguments.  Inside each one of these, I then switch on the curve type, and if it falls through, it calls a fallback function.  I don't want to put this extra logic inside each child curve type, I'd like to keep them as thin as possible.  So this is what I'm thinking:

```
Real64 CurveValue(EnergyPlusData &state, int const curveIndex, Real64 x1) {
    auto & curve = state.dataCurves->curves[curveIndex];
    switch (curve.type) {
    case CurveType::Linear:
    case CurveType::Quadratic:
    case CurveType::Cubic:
    // etc., etc.
        return curve.value();  // calls the polymorphic instance
        break;
    default:
        curve.fallback();  // alerts the user, executes the curve anyway
    }
```

It's a little clumsy to check the type when we are ultimately going to leverage polymorphism on the value() call.  But I don't see a way to avoid repeating the fallback handling.  Once we feel we've covered the corner cases, we won't need this fallback handling anymore, and the whole thing will collapse into a tiny function.  So maybe that should nudge us to hold off on the child classes until then?  Maybe not?  I'll push on the input processing and the other refactor dimensions regardless.  Thanks!

amirroth commented:
> Well, the problem is that 10 different heat pumps can share the same curve object by index. Curves have output variables for the input and output values of the curves. These were added to help debug problems, but I think they have the potential for more confusion. You have no idea which of the heat pumps was called last with its own input values and curve output. I would love to just see them go away...we should have component models register their own meaningful values as output variables, not the generic curve manager.
> 
> And on EMS, I do get it's a quick way to allow for an override, but really, we should just make sure that individual component models have the actuators they need to run so that the curve doesn't need an EMS override.
> 
> If we eliminated both of those aspects from curve manager, it would essentially become a flat set of equations. I'm not proposing we take that step now (or ever), just explaining what's going on and why.

Yes, that is the problem I meant. You have one instance of the curve which is shared by 10 components. But you have some component-level data inside the curve object so you don't know which one it belongs to. There is no real thread-safety issue if this information is not used for anything other than debugging (although it probably does make debugging more difficult) but the EMS thing is a problem, and should be pulled out of Curve and into the client object as you describe. That's an issue you always have to think about when adding this type of functionality. Material objects are shared by multiple Construction objects which are then shared by multiple Surface objects. Do you want to put X in Surface or in Construction or in Material? Which object does X have a 1-to-1 relationship with?

The debugging stuff should also probably be pulled out. You can make a struct out of it, with some debug helper functions. And then the whole thing goes away if DEBUG is not defined.

amirroth commented:
> One quick question about the child classes. Right now I've broken up the CurveValue() function into 6 overloads, each with a specific number of independent variable arguments. Inside each one of these, I then switch on the curve type, and if it falls through, it calls a fallback function. I don't want to put this extra logic inside each child curve type, I'd like to keep them as thin as possible. So this is what I'm thinking:
> 
> ```
> Real64 CurveValue(EnergyPlusData &state, int const curveIndex, Real64 x1) {
>     auto & curve = state.dataCurves->curves[curveIndex];
>     switch (curve.type) {
>     case CurveType::Linear:
>     case CurveType::Quadratic:
>     case CurveType::Cubic:
>     // etc., etc.
>         return curve.value();  // calls the polymorphic instance
>         break;
>     default:
>         curve.fallback();  // alerts the user, executes the curve anyway
>     }
> ```
> 
> It's a little clumsy to check the type when we are ultimately going to leverage polymorphism on the value() call. But I don't see a way to avoid repeating the fallback handling. Once we feel we've covered the corner cases, we won't need this fallback handling anymore, and the whole thing will collapse into a tiny function. So maybe that should nudge us to hold off on the child classes until then? Maybe not? I'll push on the input processing and the other refactor dimensions regardless. Thanks!

Well, presumably once things are polymorphic you will not actually be using this API within EnergyPlus itself, you will be using the methods. So this code will exist but only for the benefit of external APIs. Are you concerned about having fallback logic both in this function and in the method? Or having fallback handling at all? Or not being able to do fallback handling properly in the method? Or not being able to use the methods in EnergyPlus?

Myoldmopar commented:
> Are you concerned about having fallback logic both in this function and in the method? Or having fallback handling at all? Or not being able to do fallback handling properly in the method? Or not being able to use the methods in EnergyPlus?

I think I'm just going to have to start converting it to methods and monitor my blood pressure.  When I see it rise, I will assume the thing I'm changing at that point is what I am concerned about.  In this moment, nothing.

amirroth commented:
Make sure to stay hydrated. And don't forget about electrolytes!

Myoldmopar committed:
Minor cleanup in table manager

Myoldmopar committed:
Dispatching curve functionality to methods

Myoldmopar committed:
Intermediate commit

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into CurveManagerCleanDec2022

Myoldmopar commented:
OK, so I caught back up with the status of this.  I have the branch running locally with no failures or regressions.  It's definitely in an intermediate state though.
 - I haven't overhauled the input processor to use epJSON directly yet...there are lots of inputs with subtle differences in all these curve objects and it'll take some time to move cautiously through.
 - I am indeed still using the CurveValue() client API inside EnergyPlus, but it forwards the call into member methods on the curve class.  Ultimately we'll just use the member methods directly by clients, but right now this adds an extra layer. 
 
I actually think it's probably good to go in as-is assuming CI agrees with my local testing.  I'm happy to just continue forward, but it would be great to button this large set of changes up and move on with a smaller changeset.  The branch has conflicts, but I suspect it will have additional conflicts once we merge in the Optional branch and the Format branch.  So I'll aim to get those in first.

Myoldmopar committed:
Comment change

Myoldmopar committed:
Merge develop; resolve conflicts

Myoldmopar committed:
Clang-format

Myoldmopar committed:
Fix build issues on Windows

amirroth commented on src/EnergyPlus/Autosizing/CoolingCapacitySizing.cc:
```diff
@@ -189,7 +189,16 @@ Real64 CoolingCapacitySizer::size(EnergyPlusData &state, Real64 _originalValue,
+                                case 1:
+                                    TotCapTempModFac = Curve::CurveValue(state, this->dataTotCapCurveIndex, CoilInWetBulb);
+                                    break;
+                                case 2:
```
You can just put `default:` under `case 2:`.

amirroth commented on src/EnergyPlus/CurveManager.cc:
```diff
@@ -146,78 +152,441 @@ namespace Curve {
+                   this->coeff[9] * V1 * V2 * V2 + this->coeff[10] * V1 * V1 * V2 * V2 + this->coeff[11] * V3 * V2 * V2 * V2;
+        case CurveType::TriQuadratic: {
+            auto const &Tri2ndOrder(this->tri2ndOrder);
+            auto const V1s(V1 * V1);
```
:rage:

amirroth commented on src/EnergyPlus/CurveManager.cc:
```diff
@@ -146,78 +152,441 @@ namespace Curve {
+            auto const V1s(V1 * V1);
+            auto const V2s(V2 * V2);
+            auto const V3s(V3 * V3);
+            return Tri2ndOrder[0] + Tri2ndOrder[1] * V1s + Tri2ndOrder[2] * V1 + Tri2ndOrder[3] * V2s + Tri2ndOrder[4] * V2 + Tri2ndOrder[5] * V3s +
```
Why `tri2ndOrder` and not `coeff`?

Myoldmopar commented:
Still to do (a living list):
 - [x] Btwxt `find` change (https://github.com/NREL/EnergyPlus/commit/9b19c31ea0a5e8ab25b4a1296fb43b1124c62cf9#r97802937)
 - [x] Remove autos
 - [x] Change tri2ndorder to just use coeff?  
 - [x] Make sure the switch cases on the rank is set up properly -- DRY up the default: usage
 - [x] Clang format fix
 - [x] Use `std::clamp` (https://github.com/NREL/EnergyPlus/commit/e0e075d8a2a09704d0492673208133c0806ebf07#r97805625)
 - [x] Proper fix of namespace, remove quick build hack
 - [x] Use curve references/pointers in unit tests

Myoldmopar committed:
Clang format fixes

amirroth commented on src/EnergyPlus/CurveManager.hh:
```diff
@@ -328,12 +364,16 @@ struct CurveManagerData : BaseGlobalStruct
+    bool showFallbackMessage = true;
+    Array1D<Curve::Curve *> PerfCurve;
     Curve::BtwxtManager btwxtManager;
     std::unordered_map<std::string, std::string> UniqueCurveNames;
```
Can we get rid of this?

amirroth commented on src/EnergyPlus/CurveManager.hh:
```diff
@@ -328,12 +364,16 @@ struct CurveManagerData : BaseGlobalStruct
+        for (Curve::Curve *p : PerfCurve) {
+            delete p;
+        }
         *this = CurveManagerData();
```
`new (this) CurveManagerData();`

amirroth commented on src/EnergyPlus/CurveManager.hh:
```diff
@@ -328,12 +364,16 @@ struct CurveManagerData : BaseGlobalStruct
     bool FrictionFactorErrorHasOccurred = false;
-    Array1D<Curve::PerformanceCurveData> PerfCurve;
+    bool showFallbackMessage = true;
+    Array1D<Curve::Curve *> PerfCurve;
```
I would like to do something here, maybe not in this PR. I want all objects that are stored as arrays of pointers to be: i) subtypes of an EPObject class that contains one field `std::string name` (or maybe a second field `std::string nameUC`), and ii) stored in a `std::vector` and for their indices to be 0 based, with -1 meaning not found. These two things are separate of course, but would be good to knock them both out. Will do this with `Material` also.

Myoldmopar replied:
I could easily move this to an EPVector as a step in the right direction.  I didn't change it to a std::vector because the index usage is spread so far around everywhere that I was worried about index issues causing lots of diffs or failures to resolve in the same PR as other big changes.  I am happy to move forward on these ideas though as we continue the refactor.

amirroth replied:
I understand.

amirroth commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -7710,7 +7710,7 @@ void DefineCondEntSetPointManager::calculate(EnergyPlusData &state)
 
     // Using/Aliasing
-    using Curve::CurveValue;
+    using EnergyPlus::Curve::CurveValue;
```
Really?

Myoldmopar replied:
The namespace changes were just a quick fix on my Windows machine to get the build to succeed.  There were build failures due to ambiguous types that were not showing on Linux.  So that was the quickest way to get the build passing.  I'll go back through and just eliminate the `using` entirely for these cases.

amirroth commented on src/EnergyPlus/StandardRatings.cc:
```diff
@@ -563,7 +563,19 @@ namespace StandardRatings {
+                    case 3:
+                        ReformEIRChillerEIRFPLR = CurveValue(state, EIRFPLRCurveIndex, CondenserOutletTemp, reducedPLR, 0.0);
+                        break;
+                    default:
```
can put `default:` immediately after `case 3:` since they are the same case.

Myoldmopar replied:
:+1:

amirroth commented on src/EnergyPlus/StandardRatings.cc:
```diff
@@ -609,9 +621,33 @@ namespace StandardRatings {
+                    case 2:
+                        ChillerEIRFPLR = CurveValue(state, EIRFPLRCurveIndex, CondenserOutletTemp, PartLoadRatio);
+                        break;
+                    case 3:
```
Here too.

amirroth commented on src/EnergyPlus/StandardRatings.cc:
```diff
@@ -563,7 +563,19 @@ namespace StandardRatings {
                     AvailChillerCap = RefCap * ReformEIRChillerCapFT;
 
-                    ReformEIRChillerEIRFPLR = CurveValue(state, EIRFPLRCurveIndex, CondenserOutletTemp, reducedPLR);
+                    switch (state.dataCurveManager->PerfCurve(EIRFPLRCurveIndex)->numDims) {
```
You're already grabbing the curve pointer here. Any reason not to save it in an `auto *eirFPLRCurve` and use methods below?

Myoldmopar replied:
Good plan.

amirroth commented on src/EnergyPlus/StandardRatings.cc:
```diff
@@ -3097,7 +3133,7 @@ namespace StandardRatings {
 
             {
-                if (state.dataCurveManager->PerfCurve(EIRFTempCurveIndex(spnum)).numDims == 1) {
+                if (state.dataCurveManager->PerfCurve(EIRFTempCurveIndex(spnum))->numDims == 1) {
```
Better to do `auto *eirFTempCurve` and call methods rather than do the curve lookup four separate times?

amirroth commented on tst/EnergyPlus/unit/AdvancedAFN.unit.cc:
```diff
@@ -116,35 +116,36 @@ TEST_F(EnergyPlusFixture, AirflowNetwork_AdvancedTest_Test1)
-    state->dataCurveManager->PerfCurve(CurveNum).inputLimits[0].max = 10.0;
-    state->dataCurveManager->PerfCurve(CurveNum).inputLimits[1].min = 0.0;
-    state->dataCurveManager->PerfCurve(CurveNum).inputLimits[1].max = 2.0;
+    state->dataCurveManager->PerfCurve(CurveNum)->curveType = CurveType::Quadratic;
```
`auto *thisCurve`?

amirroth commented on tst/EnergyPlus/unit/AdvancedAFN.unit.cc:
```diff
@@ -116,35 +116,36 @@ TEST_F(EnergyPlusFixture, AirflowNetwork_AdvancedTest_Test1)
     state->dataCurveManager->NumCurves = 2;
-    state->dataCurveManager->PerfCurve.allocate(state->dataCurveManager->NumCurves);
+    for (int curveIndex = 1; curveIndex <= 2; curveIndex++)
+        state->dataCurveManager->PerfCurve.push_back(new EnergyPlus::Curve::Curve);
```
Are unit tests not in the `EnergyPlus` namespace?

Myoldmopar replied:
There was something odd on Windows/Visual Studio causing the namespace issue.  In this particular case, I don't think it was needed.  I can back it out and verify on Windows before pushing.

Myoldmopar committed:
Add btwxt iterator shortcuts, remove tri2ndorder array entirely and just use coeff

Myoldmopar committed:
Merge develop; resolve format conflict

Myoldmopar committed:
Fix auto usage -- leaving most of btwxt manager autos in place

Myoldmopar committed:
DRY up new rank switch blocks around the code

Myoldmopar committed:
Use std::clamp to apply min/max...super nice

Myoldmopar committed:
Remove using EnergyPlus::Curve::CurveValue, but still explicitly listing E+ namespace to build on VS

Myoldmopar committed:
Use new (this) approach in curve manager data class

Myoldmopar committed:
Use EPVector to hold curves

Myoldmopar committed:
Only look up curves once when constructing them in unit tests

Myoldmopar committed:
Clang-format

Myoldmopar commented:
OK, a bunch of things addressed, and it should be all clean.  I do intend on immediately continuing to push on the curve manager, but I would love to get this merged in so I am working with a smaller changeset.  Everyone feel free to comment on the changes here and suggest changes to make.  If they can be :pushpin: then they will be included on the next round.  Obviously if there is something actually wrong, I'll be happy to address it prior to merge.  Thanks!

amirroth commented:
I just put in a comment, but you don't need to address it before the merge. I like the pattern of function/method overloading (instead of using optional parameters) that is established in this set of changes, it's cleaner IMO and should be somewhat faster also. Lines up well with the general move away from optional arguments.

Myoldmopar commented:
Thanks @amirroth, I definitely like this as a step in the right direction.  I'm going to continue right away, but I want to touch base with you and interested teammates about some finer details to decide exactly what the final form of this looks like.

Myoldmopar commented:
Next PR steps:
 - Better client API: No more CurveValue() hanging function interface, clients store a non-owning pointer to an existing curve instance (still stored in CurveManager) and call methods on the instance.  No more index lookups.  Very much the same as PlantComponent.

...

Myoldmopar committed:
Undo the nice std::clamp changes in the CurveValue overloads

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into CurveManagerCleanDec2022

Myoldmopar commented:
OK, the clamp reversion _should_ get us back to all passing.  Shame that it had to be reverted, but we'll get to std::clamp soon enough I guess.

One more thing I want to do before this round goes in is rethink the `new` calls that are now being required in a bunch of unit test locations where the curves are being set up manually.  I think I want to make a function inside the curve manager that can isolate that manual memory management closer to where the data is being stored/used.  And actually, right now, since I haven't jumped to using polymorphism, maybe I should just back that out and store instances.  Then revisit that once we get to that point in the next round of refactor here.

amirroth commented:
> OK, the clamp reversion _should_ get us back to all passing. Shame that it had to be reverted, but we'll get to std::clamp soon enough I guess.
> 
> One more thing I want to do before this round goes in is rethink the `new` calls that are now being required in a bunch of unit test locations where the curves are being set up manually. I think I want to make a function inside the curve manager that can isolate that manual memory management closer to where the data is being stored/used. And actually, right now, since I haven't jumped to using polymorphism, maybe I should just back that out and store instances. Then revisit that once we get to that point in the next round of refactor here.

What's wrong with `new`? Seriously, what's more troubling about `new` than about `allocate()`? The fact that the destructor calls `deallocate()`? Just call `clear_state()` at the end. 

In general, I wouldn't bother with trying to hide `new` behind some pattern like "pseudo-factory". It's more important to establish a good pattern for when and how to call `delete` and `clear_state()` is a good place for that.

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into CurveManagerCleanDec2022

Myoldmopar committed:
Dont create extra state var just for octree test

Myoldmopar committed:
Add new worker method in curve state data to handle allocation

Myoldmopar commented:
OK, so I added a method called `allocateCurveVector`, which handles proper assignment of the NumCurves variable as well as proper allocation of new curve instances into the vector.  It's not getting rid of the new usage, but it keeps it out of the unit test, and honestly, I think it looks pretty nice in the unit test:

```
state->dataCurveManager->allocateCurveVector(2);
```

This is instead of repeated assignments of NumCurves followed by a loop where we are allocating the vector.  A little DRYer, and a little less likely to accidentally use a mismatched value for the number of curves you are allocating.

As I was working on this, I found a unit test that created a local `new` state instance, which it never really needed, and didn't delete it when done.  I cleaned it entirely and that unit tests runs a good 20x faster (it wasn't a time hog, but anyway...).

OK, I do believe this is going to be good to go in if CI comes back happy.  If anyone has any comments, I'm happy to address any issues today so that I can try to button this up before the weekend.

Thanks all!

amirroth commented:
> OK, so I added a method called `allocateCurveVector`, which handles proper assignment of the NumCurves variable as well as proper allocation of new curve instances into the vector. It's not getting rid of the new usage, but it keeps it out of the unit test, and honestly, I think it looks pretty nice in the unit test:
> 
> ```
> state->dataCurveManager->allocateCurveVector(2);
> ```
> 
> This is instead of repeated assignments of NumCurves followed by a loop where we are allocating the vector. A little DRYer, and a little less likely to accidentally use a mismatched value for the number of curves you are allocating.

Just want to point out that this approach will not work if `Curve` is a class hierarchy rather than a single class. We should agree on and establish a pattern for that case (and even for the single class case), but we don't have to do it as part of this PR. This PR has enough in it as is.

jasondegraw commented:
I would lose my template street cred if I didn't post this, so here goes... Once there's a class hierarchy, then something like

```
state->dataCurveManager->addCurve<Linear>();
```

would be the way to go, wouldn't it? This keeps the curve manager object in the responsible role for managing the curves, but it would require doing things a bit more incrementally (but in the unit tests that's OK, and it's definitely OK in that sketchy `makeTable` function that I've saddled you all with). I will now turn off the internet and hide.

amirroth commented:
> I would lose my template street cred if I didn't post this, so here goes... Once there's a class hierarchy, then something like
> 
> ```
> state->dataCurveManager->addCurve<Linear>();
> ```
> 
> would be the way to go, wouldn't it? This keeps the curve manager object in the responsible role for managing the curves, but it would require doing things a bit more incrementally (but in the unit tests that's OK, and it's definitely OK in that sketchy `makeTable` function that I've saddled you all with). I will now turn off the internet and hide.

Please no more new templates unless necessary. Templates shorten source code (maybe) but they duplicate machine code (definitely).

Myoldmopar commented:
> this approach will not work if `Curve` is a class hierarchy rather than a single class. We should agree on and establish a pattern for that case (and even for the single class case), but we don't have to do it as part of this PR. This PR has enough in it as is.

Understood.  Two major pieces left in the curve manager work are:
 - creating the class hierarchy, and 
 - changing the client API so that clients simply hold non-owning pointers to existing instances stored in the curve manager.  
 
Both of these will require additional thought and design, which I am excited to work on, especially isolated from the large set of changes already in this PR.

> I will now turn off the internet and hide.

lol

OK, so for today's sake, it looks like CI is all happy, just waiting on integration coverage to finish.  Assuming it's clean, I will plan on getting this merged in as-is, and immediately branching off for more work here.  Let me know if there's any reason not to do that.

Myoldmopar commented:
Alright, getting this one in, and then next up I'll work on improving the curve manager client API.  Thanks all!
